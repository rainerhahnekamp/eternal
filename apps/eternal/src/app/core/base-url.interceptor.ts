import { HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { BASE_URL } from '../shared/base-url.token';

@Injectable()
export class BaseUrlInterceptor implements HttpInterceptor {
  #baseUrl = inject(BASE_URL);

  intercept(req: HttpRequest<unknown>, next: HttpHandler): Observable<HttpEvent<unknown>> {
    if (!req.url.startsWith('/')) {
      return next.handle(req);
    }
    return next.handle(
      req.clone({
        url: `${this.#baseUrl}${req.url}`,
        withCredentials: true
      })
    );
  }
}
