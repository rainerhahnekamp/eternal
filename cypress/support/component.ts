import { mount } from 'cypress/angular';
import Chainable = Cypress.Chainable;
import '@testing-library/cypress/add-commands';

// eslint-disable-next-line @typescript-eslint/no-namespace
declare namespace Cypress {
  interface Chainable {
    mount: typeof mount;
    testid(testid: string): Chainable<JQuery>;
  }
}

Cypress.Commands.add('mount', mount);
Cypress.Commands.addQuery('testid', (testid: string) => {
  const getFn = cy.now('get', `[data-testid=${testid}]`) as () => Chainable;
  return () => getFn();
});
