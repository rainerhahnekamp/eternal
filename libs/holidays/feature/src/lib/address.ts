export interface Address {
  street: string;
  streetNumber: string;
  zip?: string;
  city?: string;
}
