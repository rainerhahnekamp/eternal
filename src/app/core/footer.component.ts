import {
  afterNextRender,
  ChangeDetectionStrategy,
  Component,
  inject,
  signal,
} from '@angular/core';
import { HeartbeatService } from '@app/heartbeat.service';
import { MatIcon } from '@angular/material/icon';

@Component({
  selector: 'app-footer',
  template: `
    <footer class="p-2 background-color-toolbar">
      @if (isHydrated()) {
        <div
          class="text-xs flex items-center justify-end gap-2"
          data-testid="hydrated"
        >
          <span>Application is ready</span>
          @if (apiReachable() === 'connected') {
            <mat-icon>wifi</mat-icon>
          } @else {
            <mat-icon class="text-red-500">wifi_off</mat-icon>
          }
        </div>
      }
    </footer>
  `,
  standalone: true,
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [MatIcon],
})
export class FooterComponent {
  readonly isHydrated = signal(false);
  readonly apiReachable = inject(HeartbeatService).status;

  constructor() {
    afterNextRender(() => this.isHydrated.set(true));
  }
}
