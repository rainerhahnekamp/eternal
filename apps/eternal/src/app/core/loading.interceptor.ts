import {
  HttpEvent,
  HttpHandler,
  HttpInterceptor,
  HttpRequest,
  HttpResponse
} from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';
import { LoadingService } from '../shared/loading.service';

@Injectable()
export class LoadingInterceptor implements HttpInterceptor {
  #loadingService = inject(LoadingService);

  intercept(req: HttpRequest<unknown>, next: HttpHandler): Observable<HttpEvent<unknown>> {
    this.#loadingService.loading();

    return next.handle(req).pipe(
      tap((event) => {
        if (event instanceof HttpResponse) {
          this.#loadingService.loaded();
        }
      })
    );
  }
}
