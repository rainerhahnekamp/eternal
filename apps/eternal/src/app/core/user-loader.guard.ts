import { inject, Injectable } from '@angular/core';
import { CanActivate } from '@angular/router';
import { Observable } from 'rxjs';
import { filter } from 'rxjs/operators';
import { SecurityService } from '../security/security.service';

@Injectable({ providedIn: 'root' })
export class UserLoaderGuard implements CanActivate {
  #securityService = inject(SecurityService);

  canActivate(): Observable<boolean> | boolean {
    return this.#securityService.loaded$.pipe(filter(Boolean));
  }
}
