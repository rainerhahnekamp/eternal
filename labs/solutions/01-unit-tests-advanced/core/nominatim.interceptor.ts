import { HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class NominatimInterceptor implements HttpInterceptor {
  intercept(req: HttpRequest<unknown>, next: HttpHandler): Observable<HttpEvent<unknown>> {
    if (req.url.match(/nominatim\.openstreetmap\.org/)) {
      return next.handle(req.clone({ setHeaders: { NominatimId: '0129' } }));
    }

    return next.handle(req);
  }
}
