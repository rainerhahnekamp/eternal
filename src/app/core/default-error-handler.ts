import { ErrorHandler, Injectable, Injector } from '@angular/core';
import { MessageService } from '../shared/ui-messaging/message/message.service';

@Injectable()
export class DefaultErrorHandler implements ErrorHandler {
  constructor(private injector: Injector) {}

  handleError(error: unknown): void {
    const messageService = this.injector.get(MessageService);
    messageService.error('We are sorry. An error happened.');
    console.error(error);
  }
}
