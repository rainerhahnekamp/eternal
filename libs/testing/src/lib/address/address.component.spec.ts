import { TestBed, ComponentFixtureAutoDetect } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { AddressComponent } from './address.component';
import { By } from '@angular/platform-browser';

describe('Test Address input', () => {
  it('should check for fields input and output', () => {
    const lookupFn = (street: string, streetNumber: string) => {
      if (street === 'Jameson Street' && streetNumber === '5') {
        return true;
      } else {
        return null;
      }
    };

    const lookup = { lookup: lookupFn };
    const component = new AddressComponent(lookup, null);
    component.search();

    expect(component.search()).toBeTruthy();
  });

  it('should check error handling when offline', () => true);

  it('should allow a user to submit an unknown address', () => true);
  it('should show multiple addresses and provide a selection', () => true);

  it('should check the snapshot', () => {
    TestBed.configureTestingModule({
      declarations: [AddressComponent],
      imports: [ReactiveFormsModule]
    });
    const fixture = TestBed.createComponent(AddressComponent);
    expect(fixture).toMatchSnapshot();
  });

  it('should check layout with input', () => {
    TestBed.configureTestingModule({
      declarations: [AddressComponent],
      imports: [ReactiveFormsModule]
    });
    const fixture = TestBed.createComponent(AddressComponent);
    const component = fixture.componentInstance;
    fixture.detectChanges();

    component.formGroup.patchValue({
      street: 'Hauptstraße',
      streetNumber: '5'
    });
    const street = fixture.debugElement.query(By.css('input[name=street]'))
      .nativeElement as HTMLInputElement;
    const streetNumber = fixture.debugElement.query(
      By.css('input[name=streetNumber]')
    ).nativeElement as HTMLInputElement;

    expect(street.value).toBe('Hauptstraße');
    expect(streetNumber.value).toBe('5');
  });

  it('should test paragraph text', () => {
    const fixture = TestBed.configureTestingModule({
      declarations: [AddressComponent],
      imports: [ReactiveFormsModule]
    }).createComponent(AddressComponent);

    fixture.detectChanges();

    const p = fixture.debugElement.query(By.css('p'))
      .nativeElement as HTMLElement;
    expect(p.textContent).toBe('Address Validation');

    fixture.componentInstance.title = 'Validator';
    fixture.detectChanges();
    expect(p.textContent).toBe('Validator');
  });
});
