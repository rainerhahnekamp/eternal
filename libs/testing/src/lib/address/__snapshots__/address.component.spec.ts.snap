// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Address input should check layout with input 1`] = `
<eternal-address
  addressLookup={[Function AddressLookup]}
  formBuilder={[Function FormBuilder]}
  formGroup={[Function FormGroup]}
>
  <form
    class="ng-untouched ng-pristine ng-valid"
    novalidate=""
  >
    <input
      class="ng-untouched ng-pristine ng-valid"
      formcontrolname="street"
    />
    <input
      class="ng-untouched ng-pristine ng-valid"
      formcontrolname="streetNumber"
    />
  </form>
</eternal-address>
`;

exports[`Test Address input should check the snapshot 1`] = `
<eternal-address
  addressLookup={[Function AddressLookup]}
  formBuilder={[Function FormBuilder]}
>
  <form
    novalidate=""
  >
    <input
      formcontrolname="from"
    />
    <input
      formcontrolname="to"
    />
  </form>
</eternal-address>
`;
